general/echo/kmdf/exe/src/main.rs

```diff
@@ -0,0 +1,489 @@
+                    globals.g_limited_loops = true;
+                } else {
+                    globals.g_limited_loops = false;
+                }
+            } else {
+                eprintln!(
+                    r##"
+Usage:
+    Echoapp.exe         --- Send single write and read request synchronously
+    Echoapp.exe -Async  --- Send reads and writes asynchronously without terminating
+    Echoapp.exe -Async <number> --- Send <number> reads and writes asynchronously
+Exit the app anytime by pressing Ctrl-C
+"##
+                );
+                return Err(String::from("Invalid Args").into());
+            }
+        }
+    }
+
+    get_device_path(&GUID_DEVINTERFACE_ECHO)?;
+
+    let globals = GLOBAL_DATA.read()?;
+    println!("DevicePath: {}", globals.g_device_path);
+
+    let h_device: HANDLE;
+    unsafe {
+        let mut path_vec = globals.g_device_path.encode_utf16().collect::<Vec<_>>();
+        path_vec.push(0);
+        let path = path_vec.as_ptr();
+
+        h_device = CreateFileW(
+            path,
+            FILE_GENERIC_READ | FILE_GENERIC_WRITE,
+            FILE_SHARE_READ | FILE_SHARE_WRITE,
+            std::ptr::null(),
+            OPEN_EXISTING,
+            0,
+            0,
+        );
+
+        if h_device == INVALID_HANDLE_VALUE {
+            return Err(
+                String::from(format!("Failed to open device. Error {}", GetLastError())).into(),
+            );
+        }
+    }
+
+    println!("Opened device successfully");
+
+    if globals.g_perform_async_io {
+        println!("Starting AsyncIo");
+
+        let h =
+            thread::spawn(|| -> Result<(), Box<dyn Error + Send + Sync>> { async_io(READER_TYPE) });
+
+        match async_io(WRITER_TYPE) {
+            Err(e) => return Err(e),
+            _ => (),
+        };
+
+        h.join().unwrap().unwrap();
+    } else {
+        perform_write_read_test(h_device, 512)?;
+
+        perform_write_read_test(h_device, 30 * 1024)?;
+    }
+
+    Ok(())
+}
+
+fn create_pattern_buffer(length: u32) -> Result<Vec<u8>, Box<dyn Error>> {
+    let mut buf = Vec::<u8>::with_capacity(length as usize);
+
+    for i in 0..length {
+        buf.push(i as u8);
+    }
+
+    Ok(buf)
+}
+
+fn verify_pattern_buffer(buf: &Vec<u8>) -> Result<(), Box<dyn Error>> {
+    for i in 0..buf.len() {
+        if buf[i] != i as u8 {
+            return Err(String::from(format!(
+                "Pattern changed.  SB 0x{:08X}, Is 0x{:08X}",
+                i as u8, buf[i]
+            ))
+            .into());
+        }
+    }
+    Ok(())
+}
+
+fn perform_write_read_test(h_device: HANDLE, test_length: u32) -> Result<(), Box<dyn Error>> {
+    let write_buffer = create_pattern_buffer(test_length)?;
+    let mut read_buffer: Vec<u8> = vec![0; test_length as usize];
+
+    let mut r: windows_sys::Win32::Foundation::BOOL;
+    let mut bytes_returned: u32 = 0;
+    unsafe {
+        r = WriteFile(
+            h_device,
+            write_buffer.as_ptr() as _,
```

Comment:
```suggestion
            write_buffer.as_ptr().cast(),
```

prefer `.cast()` to `as` for pointers to prevent unintentional changing of mutability